File: app/src/main/java/smylee/app/api/ApiClient.java
package smylee.app.api;

import com.github.simonpercic.oklog3.OkLogInterceptor;

import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import smylee.app.utils.FileUtils;

public class ApiClient {
    private static final String BASE_URL = FileUtils.MAIN_URL;
    private static Retrofit retrofit = null;

    /*private static Gson gson = new GsonBuilder()
            .setLenient()
            .create();*/

    public static Retrofit getClient() {
        // for signed app
        OkLogInterceptor okLogInterceptor = OkLogInterceptor.builder().build();
        OkHttpClient.Builder okHttpBuilder = new OkHttpClient.Builder();
        okHttpBuilder.addInterceptor(okLogInterceptor);
        OkHttpClient okHttpClient = okHttpBuilder.build();
        if (retrofit == null) {
            retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .client(okHttpClient)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }
        return retrofit;
    }

    /*public static OkHttpClient okHttpClient = new OkHttpClient.Builder()
            .readTimeout(300, TimeUnit.SECONDS)
            .connectTimeout(300, TimeUnit.SECONDS)
            .build();*/
}

File: app/src/main/java/smylee/app/api/APIClient.kt
package smylee.app.api

import APICalls
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.scalars.ScalarsConverterFactory
import smylee.app.BuildConfig
import java.util.concurrent.TimeUnit

class APIClient {

    companion object {
        private var retrofit: Retrofit? = null
        private var apiCalls: APICalls? = null

        fun getClient(): APICalls? {
            if (apiCalls == null) {
                val interceptor = HttpLoggingInterceptor()
                interceptor.level =
                    if (BuildConfig.DEBUG) HttpLoggingInterceptor.Level.BODY else HttpLoggingInterceptor.Level.NONE

                val client = OkHttpClient
                    .Builder()
                    .addInterceptor(interceptor)
                    .readTimeout(3, TimeUnit.MINUTES)
                    .connectTimeout(3, TimeUnit.MINUTES)
                    .retryOnConnectionFailure(true)
                    .build()

                retrofit = Retrofit.Builder()
                    .baseUrl(APIConst.BASE_URL)
                    .addConverterFactory(ScalarsConverterFactory.create())
                    .client(client)
                    .build()

                apiCalls = retrofit?.create(APICalls::class.java)!!
            }

            return apiCalls
        }
    }
}